src/gpu/program.rs

```diff
@@ -0,0 +1,80 @@
+use std::env;
+
+#[cfg(feature = "opencl")]
+use ec_gpu_gen::Limb64;
+use log::info;
+#[cfg(feature = "cuda")]
+use rust_gpu_tools::cuda;
+#[cfg(feature = "opencl")]
+use rust_gpu_tools::opencl;
+use rust_gpu_tools::{Device, Framework, GPUError as GpuToolsError, Program};
+
+#[cfg(not(all(feature = "cuda", feature = "opencl")))]
+use crate::gpu::error::GPUError;
+use crate::gpu::error::GPUResult;
+#[cfg(feature = "opencl")]
+use crate::gpu::sources;
+use crate::gpu::GpuEngine;
+use pairing::Engine;
+
+/// Returns the program for the preferred [`rust_gpu_tools::device::Framework`].
+///
+/// If the device supports CUDA, then CUDA is used, else OpenCL. You can force a selection with
+/// the environment variable `BELLMAN_GPU_FRAMEWORK`, which can be set either to `cuda` or `opencl`.
+pub fn program<E>(device: &Device) -> GPUResult<Program>
+where
+    E: Engine + GpuEngine,
+{
+    let framework = match env::var("BELLMAN_GPU_FRAMEWORK") {
```

Comment:
can you add a test for this? I have added a helper in `multicore` for testing env variable based behaviour

Reply:
I've moved the helper into its own module.